tracing-appender/src/rolling.rs

```diff
@@ -558,17 +566,90 @@ impl Inner {
+                format = format_description::parse("[year]-[month]-[day]-[hour]-[minute]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::HOURLY => {
+                format = format_description::parse("[year]-[month]-[day]-[hour]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::DAILY => {
+                format = format_description::parse("[year]-[month]-[day]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::NEVER => {
+                unreachable!("keep last N files feature is not available for `Rotation::NEVER`!")
+            }
+        }
+
+        let files = fs::read_dir(&self.log_directory).map(|dir| {
+            dir.filter_map(Result::ok)
+                .filter(|entry| {
+                    entry.file_name().to_string_lossy().contains(
+                        &self
+                            .log_filename_prefix
+                            .clone()
+                            .expect("prefix is always present"),
+                    )
+                })
+                .collect::<Vec<_>>()
+        });
+
+        match files {
+            Ok(files) => {
+                if files.len() >= keep_last {
+                    let mut sorted_files = vec![];
+                    for file in files {
+                        let filename = file.file_name().to_string_lossy().to_string();
+                        match filename.rfind('.') {
+                            Some(date_index) => {
+                                let date_str = filename[(date_index + 1)..].to_string();
+                                match Date::parse(&date_str, &format) {
+                                    Ok(date) => sorted_files.push((file.path(), date)),
+                                    Err(error) => eprintln!("error parsing the date: {error}"),
+                                }
+                            }
+                            None => {
+                                eprintln!("Filename do not meet the expected format!");
+                                continue;
+                            }
+                        }
+                    }
+                    sorted_files.sort_by_key(|item| item.1);
+                    // delete files, so that (n-1) files remain, because we will create another log file
+                    for (file, _) in &sorted_files[..sorted_files.len() - (keep_last - 1)] {
+                        if let Err(error) = fs::remove_file(file) {
+                            eprintln!("Failed to remove old log file: {error}");
+                        }
+                    }
+                }
+            }
+            Err(error) => {
+                eprintln!("Error reading the log directory/files: {error}")
+            }
+        }
+    }
```

Comment:
This is a style nit, not a hard blocker, but this function contains a lot of very deeply indented code, which i find kind of hard to read. I think this function would be much more readable with some judicious use of early returns instead. Here's what I might recommend (with my other suggestions applied as well):
```suggestion
        let files = fs::read_dir(&self.log_directory).map(|dir| {
            dir.filter_map(Result::ok)
                .filter(|entry| {
                    entry.file_name().to_string_lossy().contains(
                        &self
                            .log_filename_prefix
                            .clone()
                            .expect("prefix is always present"),
                    )
                })
                .collect::<Vec<_>>()
        });

        let mut files = match files {
            Ok(files) => files,
            Err(error) => {
                eprintln!("Error reading the log directory/files: {}", error);
                return;
            }
        };

        if files.len() < keep_last {
            return;
        }

        // sort the files by their creation timestamps.
        files.sort_by_key(|file| {
            file.metadata()
                .and_then(|metadata| metadata.created().map(Some))
                .unwrap_or_else(|error| {
                    eprintln!(
                        "Unable to get file creation time for {}: {}",
                        file.path().display(),
                        error
                    );
                    // if we couldn't determine the file's creation timestamp
                    // for whatever reason, it should compare as "less than"
                    // other files, so that it's not deleted.
                    None
                })
        });

        // delete files, so that (n-1) files remain, because we will create another log file
        for file in &files[..files.len() - (keep_last - 1)] {
            if let Err(error) = fs::remove_file(file.path()) {
                eprintln!(
                    "Failed to remove old log file {}: {}",
                    file.path().display(),
                    error
                );
            }
        }
```

Reply:
```
// if we couldn't determine the file's creation timestamp
// for whatever reason, it should compare as "less than"
// other files, so that it's not deleted.
```

then, won't we delete other log files, which should not be deleted, because they are the recent ones?
I get the idea of your suggestion and I agree, but I think it may create another problem

Reply:
hmm, i think maybe the best approach is to just give up on deleting any log fiiles if we can't get the creation timestamp for any of them. there's no correct way to handle that situation without potentially deleting the wrong file...

Reply:
I agree, that would be preferable 👍